core/execution/market_activity_tracker.go

```diff
@@ -171,24 +174,54 @@ func (mat *MarketActivityTracker) clipScoresAt1(scores []*types.MarketContributi
 }
 
 // MarkProposerPaid marks the proposer of the market as having been paid proposer bonus.
-func (mat *MarketActivityTracker) MarkPaidProposer(market string) {
+func (mat *MarketActivityTracker) MarkPaidProposer(market, payoutAsset string, marketsInScope []string, funder string) {
+	markets := strings.Join(marketsInScope[:], "_")
+	if len(marketsInScope) == 0 {
+		markets = "all"
+	}
+
 	if t, ok := mat.marketToTracker[market]; ok {
-		t.proposersPaid = true
+		ID := fmt.Sprintf("%s:%s:%s", payoutAsset, funder, markets)
+		if _, ok := t.proposersPaid[ID]; !ok {
+			t.proposersPaid[ID] = struct{}{}
+		}
 		mat.ss.changed = true
 	}
 }
 
-// GetEligibleProposers returns the proposer of the market is the market proposer has not been paid yet proposer bonus and the market is eligible for the bonus.
+// IsMarketEligibleForBonus returns true is the market proposer is eligible for market proposer bonus and has not been
+// paid for the combination of payout asset and marketsInScope.
 // The proposer is not market as having been paid until told to do so (if actually paid).
-func (mat *MarketActivityTracker) GetEligibleProposers(market string) []string {
+func (mat *MarketActivityTracker) IsMarketEligibleForBonus(market string, payoutAsset string, marketsInScope []string, funder string) bool {
 	t, ok := mat.marketToTracker[market]
 	if !ok {
-		return []string{}
+		return false
+	}
+
+	markets := strings.Join(marketsInScope[:], "_")
+	if len(marketsInScope) == 0 {
+		markets = "all"
+	}
+
+	marketIsInScope := false
+	if markets == "all" {
+		marketIsInScope = true
+	} else {
+		for _, v := range marketsInScope {
+			if v == market {
+				marketIsInScope = true
+				break
+			}
+		}
 	}
-	if !t.proposersPaid && mat.eligibilityChecker.IsEligibleForProposerBonus(market, t.valueTraded) {
-		return []string{t.proposer}
+
+	paid := false
+	ID := fmt.Sprintf("%s:%s:%s", payoutAsset, funder, markets)
+	if _, ok := t.proposersPaid[ID]; ok {
+		paid = true
 	}
-	return []string{}
+
+	return marketIsInScope && !paid && mat.eligibilityChecker.IsEligibleForProposerBonus(market, t.valueTraded)
```

Comment:
Can we not just return straight away is `marketIsInScope == false` instead of waiting to get all the way here?

Reply:
we can